service/frontend/templates/metered.tmpl

```diff
@@ -0,0 +1,101 @@
+import (
+	"context"
+
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/types"
+	"github.com/uber/cadence/service/frontend/api"
+	"github.com/uber/cadence/service/frontend/config"
+)
+
+{{- $nonDomainSpecificAPIs := list "Health" "DeprecateDomain" "DescribeDomain" "ListDomains" "RegisterDomain" "UpdateDomain" "GetSearchAttributes" "GetClusterInfo"}}
+{{- $domainIDAPIs := list "RecordActivityTaskHeartbeat" "RespondActivityTaskCanceled" "RespondActivityTaskCompleted" "RespondActivityTaskFailed" "RespondDecisionTaskCompleted" "RespondDecisionTaskFailed" "RespondQueryTaskCompleted"}}
+{{- $queryTaskTokenAPIs := list "RespondQueryTaskCompleted"}}
+
+{{- $interfaceName := .Interface.Name}}
+{{- $interfaceType := .Interface.Type}}
+{{- $handlerName := (index .Vars "handler")}}
+{{- $decorator := (printf "%s%s" (down $handlerName) $interfaceName) }}
+{{- $Decorator := (printf "%s%s" $handlerName $interfaceName) }}
+
+// {{$decorator}} frontend handler wrapper for authentication and authorization
+type {{$decorator}} struct {
+	handler {{.Interface.Type}}
+	logger log.Logger
+	metricsClient metrics.Client
+	domainCache cache.DomainCache
+	cfg *config.Config
+	tokenSerializer common.TaskTokenSerializer
+
+}
+
+// New{{$Decorator}} creates frontend handler with metrics and logging
+func New{{$Decorator}}(handler {{$.Interface.Type}}, logger log.Logger, metricsClient metrics.Client, domainCache cache.DomainCache, cfg *config.Config) {{.Interface.Type}} {
+	return &{{$decorator}}{
+		handler: handler,
+		logger: logger,
+		metricsClient: metricsClient,
+		domainCache: domainCache,
+		cfg: cfg,
+		tokenSerializer: common.NewJSONTaskTokenSerializer(),
+	}
+}
+
+{{range $method := .Interface.Methods}}
+func (h *{{$decorator}}) {{$method.Declaration}} {
+	{{- if eq $method.Name "Health"}}
+	{{ $method.Pass "h.handler." }}
+	{{- else}}
+	defer func() { log.CapturePanic(recover(), h.logger, &err) }()
+
+	{{- if eq $method.Name "SignalWorkflowExecution"}}
+	ctx = h.withSignalName(ctx, {{(index $method.Params 1).Name}}.GetDomain(), {{(index $method.Params 1).Name}}.GetSignalName())
+	{{- end}}
+	tags := []tag.Tag{tag.WorkflowHandlerName("{{$method.Name}}")}
+	{{- $scope := printf "metrics.Frontend%sScope" $method.Name}}
+	{{- $domainMetricTag := "metrics.DomainUnknownTag()"}}
+	{{- if not (has $method.Name $nonDomainSpecificAPIs) }}
```

Comment:
not a problem with this PR, but the fact that this there are ... kind of groupings and yet it's annoyingly hard to match them out makes me feel like we should maybe separate them out more at a struct / interface level in the future. 

this is fine, but I worry that in future templating like this will get messy beyond the pont of being readable